src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/AdaptiveChainPruner.java

```diff
@@ -29,41 +34,136 @@ public AdaptiveChainPruner(final double initialErrorProbability, final double lo
+            for (final Path<V,E> outgoingChain : vertexToGoodOutgoingChains.get(vertex)) {
+                goodChains.add(outgoingChain);
+                if(!processedVertices.contains(outgoingChain.getLastVertex())) {
+                    verticesToProcess.add(outgoingChain.getLastVertex());
+                }
+            }
 
-    }
+            for (final Path<V,E> incomingChain : vertexToGoodIncomingChains.get(vertex)) {
+                goodChains.add(incomingChain);
+                if(!processedVertices.contains(incomingChain.getFirstVertex())) {
+                    verticesToProcess.add(incomingChain.getFirstVertex());
+                }
+            }
+        }
+
+        final Set<Path<V,E>> errorChains = chains.stream().filter(c -> !goodChains.contains(c)).collect(Collectors.toSet());
+
+        // A vertex with N > 0 outgoing good chains corresponds to N - 1 variants
+        int numberOfVariantsInGraph = vertexToGoodOutgoingChains.keySet().stream()
+                .mapToInt(v -> Math.max(vertexToGoodOutgoingChains.get(v).size() - 1, 0)).sum();
+
+        if (numberOfVariantsInGraph > maxUnprunedVariants) {
+            // the vertex-to-good-incoming/outgoing chain maps contain all chains with log odds passing the threshold, even if
+            // the seeding and extending process revealed them to be errors.  We need to cull such chains for the following steps
+            for (final Path<V,E> chain : errorChains) {
+                vertexToGoodOutgoingChains.remove(chain.getFirstVertex(), chain);
+                vertexToGoodIncomingChains.remove(chain.getLastVertex(), chain);
+            }
+
+            // recalculate now that we have more accurate vertex-to-good-chains maps
+            numberOfVariantsInGraph = vertexToGoodOutgoingChains.keySet().stream()
+                    .mapToInt(v -> Math.max(vertexToGoodOutgoingChains.get(v).size() - 1, 0)).sum();
 
-    private double chainLogOdds(final Path<V,E> chain, final BaseGraph<V,E> graph, final double errorRate) {
-        if (chain.getEdges().stream().anyMatch(E::isRef)) {
-            return Double.POSITIVE_INFINITY;
+            // start with the worst good variants
+            final PriorityQueue<Path<V,E>> excessGoodChainsToPrune = new PriorityQueue<>(
+                    Comparator.comparingDouble((Path<V,E> c) -> Math.min(chainLogOdds.get(c).getLeft(), chainLogOdds.get(c).getLeft()))
+                            .thenComparing((Path<V,E> c) -> c.getFirstVertex().getSequence(), BaseUtils.BASES_COMPARATOR));
+
+            excessGoodChainsToPrune.addAll(goodChains);
+
+            while (numberOfVariantsInGraph > maxUnprunedVariants) {
+                final Path<V,E> worstGoodChain = excessGoodChainsToPrune.poll();
+                errorChains.add(worstGoodChain);
+                //if removing this chain pops a bubble, we have pruned a variant
+                if (vertexToGoodOutgoingChains.get(worstGoodChain.getFirstVertex()).size() > 1) {
```

Comment:
Hmm... I still think this code carries a risk, it seems possible to me that at this stage you could end up leaving a damaged graph by pruning good chains as part of a bubble that were not zipped together but appear in different orders. 
Imagine you have two variants A and B and they are both "good" that means 2 bubbles in the graph after puning but both variants had a second bubble that was pruned and thus they are sepearted by chaings A1, A2, B1, and B2. Now say that the order in which the lod scores shakes out the best LOD scores are A1, B1, B2, A2 and we only allow 1 maxUnprunedVariant. In this circumstance it seems possible to me that we might over prune the graph in a destructive manner. The graph will look at A2 first, remove it (wihtout decrementing numberOfVariantsInGraph), then B2 (again without decrementing numberOfVariantsInGraph) then hit B1 and remove it and arrive on the "corect" number of variants leftover in the graph (1) with the only chain left being that A1 chain. In this situation the pruning code will delete the chains A2, B1, and B2 which could have left the A1 bubble as a dangling end potentially without enough information to accurately recover it. It seems the absolute safest thing to do in this circumstance would be to "zip" the chains before calling this good chain bubble pruning in some capacity. This could either be done by reprocessing everything before pruning good chains so it will never be split across multiple chains that could be zipped or by fixing the LOD scores when you prune down to a single path.

Reply:
You're right; this is possible.  I changed the logic so that we remove all chains within a broken bubble at once.  That is, when we delete A2 we also delete A1.  Essentially we are zipping on the fly within the subgraph of chains that have not already been pruned.